Use uninit mark files during timeline init for atomic creation

SomeoneToIgnore opened the pull request:
Part of https://github.com/neondatabase/neon/pull/2239

Regular, from scratch, timeline creation involves `initdb` to be run in a separate directory, data from this directory to be imported into pageserver and, finally, timeline-related background tasks to start.

This PR ensures we don't leave behind any directories that are not marked as temporary and that pageserver removes such directories on restart, allowing timeline creation to be retried with the same IDs, if needed.

hlinnaka commented on pageserver/src/tenant.rs:
```diff
@@ -131,6 +133,257 @@ pub enum TenantState {
     Broken,
 }
 
+/// A timeline with some of its files on disk, being initialized.
+/// The struct API ensures the atomicity of the timeline init: it's either properly created and inserted into pageserver's memory, or
+/// its local files removed (at the worst case of a crash, left with a mark to get removed on the next restart).
+///
+/// A caller is responsible for proper timeline data filling before the final init.
```
```suggestion
/// A timeline with some of its files on disk, being initialized.
/// The struct API ensures the atomicity of the timeline init: it's either properly created and inserted into pageserver's memory, or
/// its local files removed (at the worst case of a crash, left with a mark to get removed on the next restart).
///
/// A caller is responsible for proper timeline data filling before the final init.
```
```suggestion
/// A timeline with some of its files on disk, being initialized.
/// This struct ensures the atomicity of the timeline init: it's either properly created and inserted into pageserver's memory, or
/// its local files are removed. In the worst case of a crash, a tombstone file is left behind, which causes the directory
/// to be removed on next restart.
///
/// The caller is responsible for proper timeline data filling before the final init.
```

SomeoneToIgnore replied:
Applied the 2nd version.

hlinnaka commented on pageserver/src/tenant.rs:
```diff
@@ -131,6 +133,257 @@ pub enum TenantState {
+                }
+                Err(e) => {
+                    should_remove_tombstone = false;
+                    error!("Failed to clean up unitialized timeline directory: {e:?}");
```
```suggestion
                    error!("Failed to clean up uninitialized timeline directory: {e:?}");
```

hlinnaka commented on libs/utils/src/crashsafe_dir.rs:
```diff
@@ -92,6 +84,44 @@ pub fn path_with_suffix_extension(original_path: impl AsRef<Path>, suffix: &str)
+    Ok(())
+}
+
+pub fn fsync(path: &Path) -> io::Result<()> {
```
This seems really verbose. What was wrong with just `File::open(path)?.sync_all()?` ? Don't the io::Errors contain enough context?

SomeoneToIgnore replied:
```rust
// in the upper mod
pub fn old_fsync(path: &Path) -> io::Result<()> {
    let file = File::open(path)?;
    file.sync_all()?;
    Ok(())
}

#[test]
    fn fsync_test() {
        println!("{:#?}", super::fsync(Path::new("bvsbsbs")));
        println!("{:#?}", super::old_fsync(Path::new("bvsbsbs")));
    }
```

and the verbose (and a bit adjusted) version will produce this output:

```
Err(
    Custom {
        kind: Other,
        error: "Failed to open the file \"bvsbsbs\": No such file or directory (os error 2)",
    },
)
Err(
    Os {
        code: 2,
        kind: NotFound,
        message: "No such file or directory",
    },
)
```

I always have issues with some io::Error and ? due to zero clue what exactly is wrong, which file is actually missing, etc.

hlinnaka commented on libs/utils/src/crashsafe_dir.rs:
```diff
@@ -92,6 +84,44 @@ pub fn path_with_suffix_extension(original_path: impl AsRef<Path>, suffix: &str)
+                .map_err(|e| {
+                    io::Error::new(
+                        io::ErrorKind::Other,
+                        format!("Failed to sync file metadata: {e}"),
```
`sync_all` flushes data and metadata. We don't know what exactly failed.

hlinnaka commented on pageserver/src/config.rs:
```diff
@@ -24,6 +25,7 @@ use crate::tenant_config::{TenantConf, TenantConfOpt};
 
 /// The name of the metadata file pageserver creates per timeline.
 pub const METADATA_FILE_NAME: &str = "metadata";
+pub const TIMELINE_TOMBSTONE_SUFFIX: &str = "___tlitmbstn";
```
How about something more readable like `_tombstone` or `_DELETE` ?

SomeoneToIgnore replied:
Have picked `_tombstone`.

hlinnaka commented on pageserver/src/tenant.rs:
```diff
@@ -131,6 +133,257 @@ pub enum TenantState {
+}
+
+impl UninitializedTimeline<'_> {
+    /// Ensures timeline data is valid, loads it into pageserver's memory and removes all temporary marks on success.
```
```suggestion
    /// Ensures timeline data is valid, loads it into pageserver's memory and removes tombstone file on success.
```

hlinnaka commented on pageserver/src/tenant.rs:
```diff
@@ -131,6 +133,257 @@ pub enum TenantState {
+}
+
+impl TimelineTombstone {
+    /// Useful for initializing timelines, existing on disk after the restart.
+    pub fn dummy() -> Self {
+        Self {
+            tombstone_deleted: true,
+            tombstone_path: PathBuf::new(),
+            timeline_path: PathBuf::new(),
+        }
+    }
```
This feels a bit weird. I think changing the tombstone field in UninitializedTimeline into `Option<TimelineTombstone>` would be more natural, and use None when there is no tombstone file.

hlinnaka replied:
Or perhaps have different structs for the different cases:
- `CreateInProgressTimeline` for a newly created branch, with tombstone to delete it on crash
- `LoadInProgressTimeline` for a timeline that exists on disk, and we're loading it. No tombstone file.

SomeoneToIgnore replied:
>  I think changing the tombstone field in UninitializedTimeline into `Option<TimelineTombstone>` would be more natural, and use None when there is no tombstone file.

I've went with the simplest approach later naturally too, and now the tombstone looks like 
```rust
#[must_use]
pub struct UninitializedTimeline<'t> {
    owning_tenant: &'t Tenant,
    timeline_id: TimelineId,
    raw_timeline: Option<(Timeline, TimelineTombstone)>,
}
```
which, presumably matches the proposal.

hlinnaka reviewed (commented):
Left some random comments. I'll continue reviewing this later

SomeoneToIgnore commented:
@hlinnaka , can you recheck the PR again?

I have found the bug, it's somewhat eerie: `insert_historic` method
https://github.com/neondatabase/neon/blob/14c623b254457f1dd4e1b0a64782cf5f568176bf/pageserver/src/tenant/layer_map.rs#L337-L343

allows to add any amount of duplicate layers (with duplicate file names) and those will be later considered "unique" in GC/compaction: https://github.com/neondatabase/neon/blob/14c623b254457f1dd4e1b0a64782cf5f568176bf/pageserver/src/tenant/timeline.rs#L1541-L1544

In my case, the bug was caused by 
```
        raw_timeline.import_timeline_from_postgres_datadir(pgdata_path, pgdata_lsn)?;
        let timeline = raw_timeline.initialize_with_lock(&mut timelines)?;
        drop(timelines);
```

code, where `raw_timeline.import_timeline_from_postgres_datadir` called `timeline.checkpoint(CheckpointConfig::Forced)` and filled the map with the layer, while `initialize_with_lock` called `load_layer_map` and added another, duplicate historic layer.

In this PR, I've changed this to have a proper sequence of actions for the timeline:
https://github.com/neondatabase/neon/pull/2489/commits/08fe22590cf71c27bdc4a34b53a0c899b337c54e

yet the underlying issue is not really fixed and a person is required to add historic layers very careful.

koivunej reviewed (commented):
Few quick comments rather than to keep these as draft while I switch back to the other PR.

SomeoneToIgnore commented on libs/utils/src/crashsafe_dir.rs:
```diff
@@ -92,6 +84,44 @@ pub fn path_with_suffix_extension(original_path: impl AsRef<Path>, suffix: &str)
+        .and_then(|file| {
+            file.sync_all()
+                .map_err(|e| {
+                    io::Error::new(
+                        io::ErrorKind::Other,
+                        format!("Failed to sync file {path:?} data and metadata: {e}"),
+                    )
```
Yes, the point is to chain those as `anyhow` does, with extra semantic information.
Anyhow looses the `ErrorKind::NotFound` case and requires some more dances around it which bloats the code, so I've decided to go with the io error.

koivunej replied:
The current version does however always return `e.kind() == ErrorKind::Other`, and the `NotFound` is lost. See: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=8515fb3fd0969be56bbc1e30d3fa582d

Perhaps the first `.map_err` is extra left over, or should use the `e.kind()`?

SomeoneToIgnore replied:
Nice catch, had used parent's kind everywhere.

SomeoneToIgnore commented on pageserver/src/tenant.rs:
```diff
@@ -1009,48 +1234,89 @@ impl Tenant {
         );
 
-        // Init temporarily repo to get bootstrap data
+        // a tombstone was placed in the parent method, nothing else can access this timeline files
```
Yes, I've inlined the code in closer to the end and forgotten to adjust the comment, fixed, thanks.

hlinnaka commented:
Should we call this something else than "tombstone"? A tombstone normally means that something was live earlier, but is now dead, and will be removed in the future. This usage is a little different. The timeline has a tombstone in the beginning, and it's removed when it's fully initialized.

Not sure what the appropriate term would be. Dead-man's switch? Poison pill?

SomeoneToIgnore commented:
"uninit mark"?

hlinnaka commented:
> "uninit mark"?

Works for me

koivunej commented on pageserver/src/tenant_mgr.rs:
```diff
@@ -265,58 +265,98 @@ fn create_tenant_files(
+
+    if creation_result.is_err() {
+        error!("Failed to create directory structure for tenant {tenant_id}, cleaning tmp data");
+        if let Err(e) = fs::remove_dir_all(&temporary_tenant_dir) {
```
`crashsafe_dir::create_dir_all` is used above and the renames done by `try_create_target_tenant_dir` after writing the config file are fsynced but the removal is not crashsafe. Should there be at least one fsync here in the err path?

SomeoneToIgnore replied:
It won't hurt, added, thank you.

koivunej commented on pageserver/src/tenant.rs:
```diff
@@ -119,6 +121,234 @@ pub struct Tenant {
+                    "Removing intermediate uninit mark file {}",
+                    self.uninit_mark_path.display()
+                );
+                if let Err(e) = fs::remove_file(&self.uninit_mark_path) {
+                    error!("Failed to remove the uninit mark file: {e}")
+                }
```
Similar question about fsync after remove as elsewhere, which can be found on the `TimelineUninitMark::remove_uninit_mark`. I wonder if the same piece of code to delete could be called from both `drop` and `remove_uninit_mark`.

SomeoneToIgnore replied:
Indeed, feels like the most of the deletion could be deduplicated, done.

koivunej reviewed (commented):
Few more comments, mostly concerning fsync after remove or exists+create_truncate. There's a lot of changes in parts I'm mostly seeing for the first time, so unsure if I should approve.

SomeoneToIgnore commented on pageserver/src/tenant.rs:
```diff
@@ -1063,34 +1341,77 @@ impl Tenant {
+            timeline_path.display()
+        );
+
+        fs::File::create(&uninit_mark_path)
```
I was trying to be on the safe here, when one of the pageserver parts exits, leaving the unint mark, but it does not matter with this code, since `fs::File::create(&uninit_mark_path)` will fail if the file exists.
Removed the existence check.

koivunej replied:
Note: `std::fs::File` explicitly documents that it doesn't fail if the file exists, it truncates: https://doc.rust-lang.org/stable/std/fs/struct.File.html#method.create

I think you are assuming it to be `OpenOptions::create_new` but it's not.

SomeoneToIgnore replied:
So, back to your initial comment, implemented. 
I think it's ok generally, since besides `uninit_mark_path`, we could only leave tmp file and tmp initdb dir created.
The latter gets removed if init is attempted again, the former is overwritten, so should be ok.

koivunej replied:
> The latter gets removed if init is attempted again, the former is overwritten, so should be ok.

I am not sure if I can see this right away even from the review, spent some time yesterday looking for it. Resolving as I am agreeing that it most likely doesn't matter with this code.

LizardWizzard commented on pageserver/src/tenant.rs:
```diff
@@ -1063,34 +1344,71 @@ impl Tenant {
-            .cloned();
         let new_timeline = self
-            .initialize_new_timeline(new_timeline_id, new_metadata, ancestor)
+            .create_timeline_data(new_timeline_id, new_metadata, ancestor)
```
If error appears here, then uninit mark is dropped, and timeline will be considered healthy during restart? Am I correct?

SomeoneToIgnore replied:
> I would imagine that timeline is completely initialized in temporary location (with all initdb loading) and then moved to permatent location with a rename.

That is a great idea and I've tried doing it, but the issue is that we have to write into the tmp directory during initdb result imports. 
That would require is to pass the knowledge about the tmp dir from `Tenant` level to `DeltaLayer` (and others') level: right now, we import the data and do checkpointing. During checkpointing, we use `conf.rs` functions to get the timeline path and derive the layer path to write into out of it. 
Changing that would require a lot of code, unless I'm mistaken somehow.

So, that's why the whole tombstone approach as a "simpler" way of doing things.

SomeoneToIgnore replied:
> If error appears here, then uninit mark is dropped, and timeline will be considered healthy during restart? Am I correct?

I don't think so: with the first step we create the timeline dir, then we would check that in the mark drop logic and keep the mark.
But you're right that we can make it better and at least try to remove the files right away, if fail to create the timeline: if we succeed, we could try to create this timeline before the next restart.

I've moved the `after-timeline-uninit-mark-creation` failpoint to this code and adjusted the corresponding test to not to restart pagerver at all.

LizardWizzard replied:
> then we would check that in the mark drop logic and keep the mark.

You're right. I've missed it. To me tombstone approach is way harder to reason about. Indeed many changes are required to pass around different base directory. Not sure exactly how many, I havent tried it. Hard to say is it worth it to make more changes but make it easier (subjectively) to reason about it

LizardWizzard replied:
Also coverage thinks that this branch in drop was not covered:
https://neondatabase.github.io/zenith-coverage-data/02d93d71a0ee9b43225ad4dc15613106ff62c4c2/coverage/pageserver/src/tenant.rs.html#L315

Is it coverage problem or test doesnt trigger this case?

SomeoneToIgnore replied:
Neither, it's a special handle: current PR logic actually tries to remove the timeline dir first before removing the mark, and only if the timeline dir removal fails the mark code that you've linked would be triggered.
Not sure how realistic this is, but decided it's safer to keep the mark aware of this.

SomeoneToIgnore replied:
Timelines' init tmp dir approach issue: https://github.com/neondatabase/neon/issues/2670

SomeoneToIgnore commented on libs/utils/src/crashsafe_dir.rs:
```diff
@@ -92,6 +84,36 @@ pub fn path_with_suffix_extension(original_path: impl AsRef<Path>, suffix: &str)
         .with_extension(new_extension.as_ref())
 }
 
+pub fn fsync_file_and_parent(file_path: &Path) -> io::Result<()> {
```
Renamed.

SomeoneToIgnore commented on pageserver/src/tenant.rs:
```diff
@@ -119,6 +121,237 @@ pub struct Tenant {
+                            )
+                        })?;
+                }
+                new_timeline.launch_wal_receiver().with_context(|| {
```
Good idea, have moved `launch_wal_receiver` below everything and will wait for your patch to land, thanks for noticing.

SomeoneToIgnore commented on pageserver/src/tenant.rs:
```diff
@@ -119,6 +121,237 @@ pub struct Tenant {
+
+    fn initialize_with_lock(
+        mut self,
+        timelines: &mut MutexGuard<HashMap<TimelineId, Arc<Timeline>>>,
```
Mostly semantics: adding something into &mut map is more awkward and we have to lock to get that &mut anyway.
But changed that, since no strong opinion here.

SomeoneToIgnore commented on libs/utils/src/crashsafe_dir.rs:
```diff
@@ -92,6 +84,36 @@ pub fn path_with_suffix_extension(original_path: impl AsRef<Path>, suffix: &str)
+                .map_err(|e| {
+                    io::Error::new(e.kind(), format!("Failed to fsync file {path:?}: {e}"))
+                })
+        })
```
Actually, the latter, thank you for spotting, fixed.

LizardWizzard reviewed (commented):
Logic around uninit mark looks correct to me, but I find it quite hard to reason about it.

I wonder is it worth it for me to try implementing another approach with one temp dir? It looks like this approach can be quite generic, something like `ScratchDir` that removes itself on drop, and path to it is passed as a base for tenant creation function and timeline creation with initial checkpoint. First doesnt sound too complicated, but dealing with checkpoint to some temporary location indeed looks problematic, the main problem I see is that we need to adjust path for a layer after final rename to target directory (the `path_or_conf` field). Since you actually tried it most probably it is not worth it to repeat your way

Any other thoughts about that?

SomeoneToIgnore commented on pageserver/src/tenant.rs:
```diff
@@ -119,6 +121,216 @@ pub struct Tenant {
+        let new_timeline = Arc::new(new_timeline);
+
+        let new_disk_consistent_lsn = new_timeline.get_disk_consistent_lsn();
+        // TODO it would be good to ensure that, but apparently a lot of our testing is dependend on that at least
```
All the unit tests in the same module, below.

SomeoneToIgnore commented:
> Since you actually tried it most probably it is not worth it to repeat your way

No other good thoughts, but I could be wrong about it and somebody else might try to take a stab at it, at least to see if it's all as dire as I see it.

LizardWizzard reviewed (approved):
This PR is still an improvement, I'll try to find a simpler way with temporary directory, but dont think it is a good idea to block this PR on that. Not sure when I'll get to that so +1 from me

hlinnaka commented:
Regression tests are printing new warnings in the logs with this:

```
2022-10-21T08:28:22.955233Z ERROR tenant_create{tenant=4caab9ae07b33d796c4495dff4cb8b37}: Failed to fsync removed temporary tenant directory "/tmp/test_output/test_tenant_creation_fails/repo/tenants/4caab9ae07b33d796c4495dff4cb8b37.___temp": Failed to fsync file "/tmp/test_output/test_tenant_creation_fails/repo/tenants/4caab9ae07b33d796c4495dff4cb8b37.___temp": Failed to open the file "/tmp/test_output/test_tenant_creation_fails/repo/tenants/4caab9ae07b33d796c4495dff4cb8b37.___temp": No such file or directory (os error 2)
```
and
```
2022-10-21T08:20:48.973050Z  WARN timeline_create{tenant=8476a13bcdfa964c806fd34b2c675ac1 new_timeline=None lsn=Some(0/2D469A8) pg_version=None}: Removing intermediate uninit mark file /tmp/test_output/test_branch_creation_before_gc/repo/tenants/8476a13bcdfa964c806fd34b2c675ac1/timelines/ea298ed3e567befaf0ee73611a0e2ed7.___uninit
```

Are these expected?

I'm seeing these now that I rebased PR #2653 over this:
https://neon-github-public-dev.s3.amazonaws.com/reports/pr-2653/debug/3295819192/index.html#suites/ec4311502db344eee91f1354e9dc839b/141e19d82f96d89a/.

SomeoneToIgnore commented:
I think they are normal, because one way or another trigger timeline creation failure: in this case, we don't preserve the tmp directory and remove the fallpoint.

In the suite linked:

*  `test_branch_creation_before_gc` and `test_branch_behind` fail to create timeline due to `Timeline::check_lsn_is_in_scope` failure and drops the mark which results in the logs
* `test_tenant_creation_fails`, `test_timeline_create_break_after_uninit_mark` and `test_timeline_init_break_before_checkpoint` are the tests that break timelines and tenants, so they expect to produce such logs.

Other couple of test failing in that suite seem to have other error logs, not related to the uninit marks, so all seems to be ok for me.
